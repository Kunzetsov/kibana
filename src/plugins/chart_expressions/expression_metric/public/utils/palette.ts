/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { isColorDark } from '@elastic/eui';
import { Datatable } from 'src/plugins/expressions';

export type ColumnGroups<T = number[]> = Record<string, T>;

export interface MinMax {
  min: number;
  max: number;
}

export const getMinMax = (elements: number[]) => {
  if (!elements.length) {
    return { min: -Infinity, max: -Infinity };
  }

  const sortedArray = elements.sort((a, b) => a - b);
  return { min: sortedArray[0], max: sortedArray[sortedArray.length - 1] };
};

export const groupValuesByColumns = (datatable: Datatable) => {
  const numericColumns = datatable.columns
    .filter((column) => column.meta.type === 'number')
    .map((column) => column.id);

  return numericColumns.reduce<ColumnGroups>((groups, columnId) => {
    groups[columnId] = datatable.rows.map((row) => row[columnId]);
    return groups;
  }, {});
};

export const getMinMaxForColumns = (datatable: Datatable) => {
  const groups = groupValuesByColumns(datatable);
  return Object.keys(groups).reduce<ColumnGroups<MinMax>>((minMaxGroups, columnId) => {
    minMaxGroups[columnId] = getMinMax(groups[columnId]);
    return minMaxGroups;
  }, {});
};

export const parseRgbString = (rgb: string) => {
  const [group, r, g, b, opacityGroup, opacity] =
    rgb.match(/rgb\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*?(,\s*(\d+)\s*)?\)/) ?? [];

  return {
    red: parseFloat(r),
    green: parseFloat(g),
    blue: parseFloat(b),
    opacity: opacity ? parseFloat(opacity) : undefined,
  };
};

export const shouldApplyColor = (color: string) => {
  const { opacity } = parseRgbString(color);
  return opacity !== 0; // if opacity === 0 it means, there is no color to apply to the metric
};

export const needsLightText = (bgColor: string = '') => {
  const { red, green, blue, opacity } = parseRgbString(bgColor);
  return isColorDark(red, green, blue) && opacity !== 0;
};
