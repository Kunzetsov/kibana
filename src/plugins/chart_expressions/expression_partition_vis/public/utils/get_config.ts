/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { PartitionConfig, PartitionLayout, RecursivePartial, Theme } from '@elastic/charts';
import { Datatable } from '../../../../../../src/plugins/expressions';
import {
  ChartTypes,
  LabelPositions,
  PartitionVisParams,
  PieContainerDimensions,
} from '../../common/types';

type Config = RecursivePartial<PartitionConfig>;

type GetConfigByTypeFn = (
  chartType: ChartTypes,
  visParams: PartitionVisParams,
  visData: Datatable,
  chartTheme: RecursivePartial<Theme>,
  dimensions?: PieContainerDimensions,
  rescaleFactor?: number
) => Config;

type GetConfigFn = (
  visParams: PartitionVisParams,
  visData: Datatable,
  chartTheme: RecursivePartial<Theme>,
  dimensions?: PieContainerDimensions,
  rescaleFactor?: number
) => Config;

const MAX_SIZE = 1000;

const getPieDonutWaffleCommonConfig: GetConfigFn = (
  visParams,
  visData,
  chartTheme,
  dimensions,
  rescaleFactor = 1
) => {
  const isSplitChart = Boolean(visParams.dimensions.splitColumn || visParams.dimensions.splitRow);
  const preventLinksFromShowing =
    (visParams.labels.position === LabelPositions.INSIDE || isSplitChart) && visParams.labels.show;
  const saveRescaleFactor = visParams.labels.show && !preventLinksFromShowing;

  const usingOuterSizeRatio =
    dimensions && !isSplitChart && !saveRescaleFactor
      ? {
          outerSizeRatio:
            // Cap the ratio to 1 and then rescale
            rescaleFactor * Math.min(MAX_SIZE / Math.min(dimensions?.width, dimensions?.height), 1),
        }
      : {};

  const config: Config = { ...usingOuterSizeRatio };

  if (!visParams.labels.show || visParams.labels.position === LabelPositions.HIDE) {
    // Force all labels to be linked, then prevent links from showing
    config.linkLabel = { maxCount: 0, maximumSection: Number.POSITIVE_INFINITY };
  }

  if (visParams.labels.last_level && visParams.labels.show) {
    config.linkLabel = {
      maxCount: Number.POSITIVE_INFINITY,
      maximumSection: Number.POSITIVE_INFINITY,
      maxTextLength: visParams.labels.truncate ?? undefined,
    };
  }

  if (preventLinksFromShowing) {
    config.linkLabel = { maxCount: 0 };
  }

  if (saveRescaleFactor) {
    // shrink up to 20% to give some room for the linked values
    config.outerSizeRatio = rescaleFactor;
  }

  return { ...config };
};

const getTreemapMosaicCommonConfig: GetConfigFn = (visParams) => {
  if (!visParams.labels.show || visParams.labels.position === LabelPositions.HIDE) {
    return {
      fillLabel: { textColor: 'rgba(0,0,0,0)' },
    };
  }
  return {};
};

const getPieSpecificConfig: GetConfigFn = (...args) => {
  return {
    partitionLayout: PartitionLayout.sunburst,
    ...getPieDonutWaffleCommonConfig(...args),
  };
};

const getDonutSpecificConfig: GetConfigFn = (visParams, ...args) => {
  return {
    partitionLayout: PartitionLayout.sunburst,
    emptySizeRatio: visParams.emptySizeRatio,
    ...getPieDonutWaffleCommonConfig(visParams, ...args),
  };
};

const getWaffleSpecificConfig: GetConfigFn = (...args) => {
  return {
    partitionLayout: PartitionLayout.waffle,
    ...getPieDonutWaffleCommonConfig(...args),
  };
};

const getTreemapSpecificConfig: GetConfigFn = (...args) => {
  return {
    partitionLayout: PartitionLayout.treemap,
    ...getTreemapMosaicCommonConfig(...args),
  };
};

const getMosaicSpecificConfig: GetConfigFn = (...args) => {
  return {
    partitionLayout: PartitionLayout.mosaic,
    ...getTreemapMosaicCommonConfig(...args),
  };
};

const getSpecificConfig: GetConfigByTypeFn = (chartType, ...args) =>
  ({
    [ChartTypes.PIE]: getPieSpecificConfig(...args),
    [ChartTypes.DONUT]: getDonutSpecificConfig(...args),
    [ChartTypes.TREEMAP]: getTreemapSpecificConfig(...args),
    [ChartTypes.MOSAIC]: getMosaicSpecificConfig(...args),
    [ChartTypes.WAFFLE]: getWaffleSpecificConfig(...args),
  }[chartType]);

export const getConfig: GetConfigByTypeFn = (
  chartType,
  visParams,
  visData,
  chartTheme,
  dimensions,
  rescaleFactor = 1
) => {
  // On small multiples we want the labels to only appear inside
  const isSplitChart = Boolean(visParams.dimensions.splitColumn || visParams.dimensions.splitRow);
  const usingMargin =
    dimensions && !isSplitChart
      ? {
          margin: {
            top: (1 - Math.min(1, MAX_SIZE / dimensions?.height)) / 2,
            bottom: (1 - Math.min(1, MAX_SIZE / dimensions?.height)) / 2,
            left: (1 - Math.min(1, MAX_SIZE / dimensions?.width)) / 2,
            right: (1 - Math.min(1, MAX_SIZE / dimensions?.width)) / 2,
          },
        }
      : null;
  return {
    fontFamily: chartTheme.barSeriesStyle?.displayValue?.fontFamily,
    outerSizeRatio: 1,
    specialFirstInnermostSector: false, // ?
    minFontSize: 10,
    maxFontSize: 16,
    emptySizeRatio: 0,
    sectorLineStroke: chartTheme.lineSeriesStyle?.point?.fill,
    sectorLineWidth: 1.5,
    circlePadding: 4,
    linkLabel: {
      maxCount: 5,
      fontSize: 11,
      textColor: chartTheme.axes?.axisTitle?.fill,
      maxTextLength: visParams.labels.truncate ?? undefined,
    },
    ...usingMargin,
    ...getSpecificConfig(chartType, visParams, visData, chartTheme, dimensions, rescaleFactor),
  };
};
