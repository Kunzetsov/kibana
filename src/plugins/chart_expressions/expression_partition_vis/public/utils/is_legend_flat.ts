/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { DatatableColumn } from '../../../../expressions';
import { ChartTypes } from '../../common/types';

type GetLegendIsFlatFn = (splitChartDimension: DatatableColumn | undefined) => boolean;

const isLegendFlatCommon: GetLegendIsFlatFn = (splitChartDimension) => Boolean(splitChartDimension);

const isLegendFlatWaffle: GetLegendIsFlatFn = (splitChartDimension) => true;

export const isLegendFlat = (
  visType: ChartTypes,
  splitChartDimension: DatatableColumn | undefined
) =>
  ({
    [ChartTypes.PIE]: isLegendFlatCommon(splitChartDimension),
    [ChartTypes.DONUT]: isLegendFlatCommon(splitChartDimension),
    [ChartTypes.TREEMAP]: isLegendFlatCommon(splitChartDimension),
    [ChartTypes.MOSAIC]: isLegendFlatCommon(splitChartDimension),
    [ChartTypes.WAFFLE]: isLegendFlatWaffle(splitChartDimension),
  }[visType]);
