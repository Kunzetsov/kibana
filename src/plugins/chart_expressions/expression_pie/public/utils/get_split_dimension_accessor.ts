/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
import { AccessorFn } from '@elastic/charts';
import { getColumnByAccessor } from './accessor';
import { DatatableColumn } from '../../../../expressions/public';
import type { FieldFormatsStart } from '../../../../field_formats/public';
import { Dimension } from '../../common/types';

export const getSplitDimensionAccessor =
  (fieldFormats: FieldFormatsStart, columns: DatatableColumn[]) =>
  (splitDimension: Dimension): AccessorFn => {
    const formatter = fieldFormats.deserialize(splitDimension.format);
    const splitChartColumn = getColumnByAccessor(splitDimension.accessor, columns);
    const { id } = splitChartColumn ?? {};

    const fn: AccessorFn = (d) => {
      const v = d[accessor];
      if (v === undefined) {
        return;
      }
      const f = formatter.convert(v);
      return f;
    };

    return fn;
  };
