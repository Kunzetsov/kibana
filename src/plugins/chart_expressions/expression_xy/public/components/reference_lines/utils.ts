/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Position } from '@elastic/charts';
import { IconPosition, YAxisMode } from '../../../common/types';

// if there's just one axis, put it on the other one
// otherwise use the same axis
// this function assume the chart is vertical
export function getBaseIconPlacement(
  iconPosition: IconPosition | undefined,
  axesMap?: Record<string, unknown>,
  axisMode?: YAxisMode
) {
  if (iconPosition === 'auto') {
    if (axisMode === 'bottom') {
      return Position.Top;
    }
    if (axesMap) {
      if (axisMode === 'left') {
        return axesMap.right ? Position.Left : Position.Right;
      }
      return axesMap.left ? Position.Right : Position.Left;
    }
  }

  if (iconPosition === 'left') {
    return Position.Left;
  }
  if (iconPosition === 'right') {
    return Position.Right;
  }
  if (iconPosition === 'below') {
    return Position.Bottom;
  }
  return Position.Top;
}
