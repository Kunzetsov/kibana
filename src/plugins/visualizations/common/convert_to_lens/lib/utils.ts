/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { isEqual, omit } from 'lodash';
import { IAggConfig, METRIC_TYPES } from '@kbn/data-plugin/common';
import { DataViewField } from '@kbn/data-views-plugin/common';
import { DataViewFieldBase } from '@kbn/es-query';
import { BaseSchemaConfig } from '../../types';
import { Column } from '../types';
import {
  MetricsWithoutSpecialParams,
  Column as ColumnWithMeta,
  SiblingPipelineMetric,
} from './convert';

type UnwrapArray<T> = T extends Array<infer P> ? P : T;

export const getLabel = (agg: BaseSchemaConfig) => {
  return agg.aggParams && 'customLabel' in agg.aggParams
    ? agg.aggParams.customLabel ?? agg.label
    : agg.label;
};

export const getLabelForPercentile = (agg: BaseSchemaConfig) => {
  return agg.aggParams && 'customLabel' in agg.aggParams && agg.aggParams.customLabel !== ''
    ? agg.label
    : '';
};

export const getValidColumns = (columns: Array<Column | null> | Column | null | undefined) => {
  if (columns && Array.isArray(columns)) {
    const nonNullColumns = columns.filter(
      (c): c is Exclude<UnwrapArray<typeof columns>, null> => c !== null
    );

    if (nonNullColumns.length !== columns.length) {
      return null;
    }

    return nonNullColumns;
  }

  return columns ? [columns] : null;
};

export const getFieldNameFromField = (
  field: DataViewField | DataViewFieldBase | string | undefined
) => {
  if (!field) {
    return null;
  }

  if (typeof field === 'string') {
    return field;
  }

  return field.name;
};

export const isSchemaConfig = (agg: BaseSchemaConfig | IAggConfig): agg is BaseSchemaConfig => {
  if ((agg as BaseSchemaConfig).aggType) {
    return true;
  }
  return false;
};

export const isColumnWithMeta = (column: Column): column is ColumnWithMeta => {
  if ((column as ColumnWithMeta).meta) {
    return true;
  }
  return false;
};

const SIBBLING_PIPELINE_AGGS: string[] = [
  METRIC_TYPES.AVG_BUCKET,
  METRIC_TYPES.SUM_BUCKET,
  METRIC_TYPES.MAX_BUCKET,
  METRIC_TYPES.MIN_BUCKET,
];

const PARENT_PIPELINE_AGGS: string[] = [
  METRIC_TYPES.CUMULATIVE_SUM,
  METRIC_TYPES.DERIVATIVE,
  METRIC_TYPES.MOVING_FN,
];

const AGGS_WITHOUT_SPECIAL_RARAMS: string[] = [
  METRIC_TYPES.AVG,
  METRIC_TYPES.COUNT,
  METRIC_TYPES.MAX,
  METRIC_TYPES.MIN,
  METRIC_TYPES.SUM,
  METRIC_TYPES.MEDIAN,
  METRIC_TYPES.CARDINALITY,
  METRIC_TYPES.VALUE_COUNT,
];

const PIPELINE_AGGS: string[] = [...SIBBLING_PIPELINE_AGGS, ...PARENT_PIPELINE_AGGS];

export const isSiblingPipeline = (
  metric: BaseSchemaConfig
): metric is BaseSchemaConfig<SiblingPipelineMetric> => {
  return SIBBLING_PIPELINE_AGGS.includes(metric.aggType);
};

export const isPipeline = (
  metric: BaseSchemaConfig
): metric is
  | BaseSchemaConfig<METRIC_TYPES.CUMULATIVE_SUM>
  | BaseSchemaConfig<METRIC_TYPES.DERIVATIVE>
  | BaseSchemaConfig<METRIC_TYPES.MOVING_FN>
  | BaseSchemaConfig<METRIC_TYPES.AVG_BUCKET>
  | BaseSchemaConfig<METRIC_TYPES.MAX_BUCKET>
  | BaseSchemaConfig<METRIC_TYPES.MIN_BUCKET>
  | BaseSchemaConfig<METRIC_TYPES.SUM_BUCKET> => {
  return PIPELINE_AGGS.includes(metric.aggType);
};

export const isMetricAggWithoutParams = (
  metric: BaseSchemaConfig
): metric is BaseSchemaConfig<MetricsWithoutSpecialParams> => {
  return AGGS_WITHOUT_SPECIAL_RARAMS.includes(metric.aggType);
};

export const isPercentileAgg = (
  metric: BaseSchemaConfig
): metric is BaseSchemaConfig<METRIC_TYPES.PERCENTILES> => {
  return metric.aggType === METRIC_TYPES.PERCENTILES;
};

export const isPercentileRankAgg = (
  metric: BaseSchemaConfig
): metric is BaseSchemaConfig<METRIC_TYPES.PERCENTILE_RANKS> => {
  return metric.aggType === METRIC_TYPES.PERCENTILE_RANKS;
};

export const isStdDevAgg = (
  metric: BaseSchemaConfig
): metric is BaseSchemaConfig<METRIC_TYPES.STD_DEV> => {
  return metric.aggType === METRIC_TYPES.STD_DEV;
};

export const getCutomBucketsFromSiblingAggs = (metrics: BaseSchemaConfig[]) => {
  return metrics.reduce<IAggConfig[]>((acc, metric) => {
    if (
      isSiblingPipeline(metric) &&
      metric.aggParams?.customBucket &&
      acc.every(
        (bucket) =>
          !isEqual(
            omit(metric.aggParams?.customBucket?.serialize(), ['id']),
            omit(bucket.serialize(), ['id'])
          )
      )
    ) {
      acc.push(metric.aggParams.customBucket);
    }

    return acc;
  }, []);
};
