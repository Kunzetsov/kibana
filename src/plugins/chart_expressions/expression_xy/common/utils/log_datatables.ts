/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { ExecutionContext } from '../../../../expressions';
import { prepareLogTable } from '../../../../visualizations/common/utils';
import { LayerTypes } from '../constants';
import { strings } from '../i18n';
import { CommonXYLayerConfigResult } from '../types';

export const logDatatables = (layers: CommonXYLayerConfigResult[], handlers: ExecutionContext) => {
  if (!handlers?.inspectorAdapters?.tables) {
    return;
  }

  layers.forEach((layer, index) => {
    if (layer.layerType === LayerTypes.ANNOTATIONS) {
      return;
    }

    let xAccessor;
    let splitAccessor;
    if (layer.layerType === LayerTypes.DATA) {
      xAccessor = layer.xAccessor;
      splitAccessor = layer.splitAccessor;
    }

    const { accessors, layerType } = layer;
    const logTable = prepareLogTable(
      layer.table,
      [
        [
          accessors ? accessors : undefined,
          layerType === LayerTypes.DATA ? strings.getMetricHelp() : strings.getReferenceLineHelp(),
        ],
        [xAccessor ? [xAccessor] : undefined, strings.getXAxisHelp()],
        [splitAccessor ? [splitAccessor] : undefined, strings.getBreakdownHelp()],
      ],
      true
    );

    handlers.inspectorAdapters.tables.logDatatable(index, logTable);
  });
};
