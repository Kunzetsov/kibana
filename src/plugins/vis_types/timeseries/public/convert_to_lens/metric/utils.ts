/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
import { Layer } from '@kbn/visualizations-plugin/common/convert_to_lens';
import { Layer as ExtendedLayer, excludeMetaFromColumn } from '../lib/convert';

export const excludeMetaFromLayers = (
  layers: Record<string, ExtendedLayer>
): Record<string, Layer> => {
  const newLayers: Record<string, Layer> = {};
  Object.entries(layers).forEach(([layerId, layer]) => {
    const columns = layer.columns.map(excludeMetaFromColumn);
    newLayers[layerId] = { ...layer, columns };
  });

  return newLayers;
};
