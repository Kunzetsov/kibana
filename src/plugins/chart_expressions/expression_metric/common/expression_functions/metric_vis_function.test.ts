/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { metricVisFunction } from './metric_vis_function';
import type { MetricArguments } from '../../common';
import { functionWrapper } from '../../../../expressions/common/expression_functions/specs/tests/utils';
import { Datatable } from '../../../../expressions/common/expression_types/specs';

describe('interpreter/functions#metric', () => {
  const fn = functionWrapper(metricVisFunction());
  const context = {
    type: 'datatable',
    rows: [{ 'col-0-1': 0 }],
    columns: [{ id: 'col-0-1', name: 'Count' }],
  } as unknown as Datatable;
  const args = {
    percentageMode: false,
    colorSchema: 'Green to Red',
    metricColorMode: 'None',
    palette: {
      params: {
        colors: ['rgb(0, 0, 0, 0)', 'rgb(112, 38, 231)'],
        stops: [0, 10000],
        gradient: false,
        rangeMin: 0,
        rangeMax: 150,
        range: 'number',
      },
    },
    labels: {
      show: true,
    },
    invertColors: false,
    style: {
      bgFill: '',
      bgColor: false,
      labelColor: false,
      subText: '',
      fontSize: 60,
    },
    font: { spec: { fontSize: 60 } },
    metrics: [
      {
        accessor: 0,
        format: {
          id: 'number',
        },
        params: {},
        aggType: 'count',
      },
    ],
  } as unknown as MetricArguments;

  it('returns an object with the correct structure', () => {
    const actual = fn(context, args, undefined);

    expect(actual).toMatchSnapshot();
  });

  it('logs correct datatable to inspector', async () => {
    let loggedTable: Datatable;
    const handlers = {
      inspectorAdapters: {
        tables: {
          logDatatable: (name: string, datatable: Datatable) => {
            loggedTable = datatable;
          },
        },
      },
    };
    await fn(context, args, handlers as any);

    expect(loggedTable!).toMatchSnapshot();
  });
});
