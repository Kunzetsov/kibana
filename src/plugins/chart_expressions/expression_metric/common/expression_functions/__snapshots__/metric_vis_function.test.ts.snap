// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interpreter/functions#metric logs correct datatable to inspector 1`] = `
Object {
  "columns": Array [
    Object {
      "id": "col-0-1",
      "meta": Object {
        "dimensionName": undefined,
      },
      "name": "Count",
    },
  ],
  "rows": Array [
    Object {
      "col-0-1": 0,
    },
  ],
  "type": "datatable",
}
`;

exports[`interpreter/functions#metric returns an object with the correct structure 1`] = `
Object {
  "as": "metricVis",
  "type": "render",
  "value": Object {
    "visConfig": Object {
      "dimensions": Object {
        "metrics": undefined,
      },
      "metric": Object {
        "labels": Object {
          "show": true,
        },
        "metricColorMode": "\\"None\\"",
        "palette": Object {
          "colors": Array [
            "rgb(0, 0, 0, 0)",
            "rgb(112, 38, 231)",
          ],
          "gradient": false,
          "range": "number",
          "rangeMax": 150,
          "rangeMin": 0,
          "stops": Array [
            0,
            10000,
          ],
        },
        "percentageMode": false,
        "style": Object {
          "bgColor": false,
          "bgFill": "",
          "fontSize": 60,
          "labelColor": false,
          "subText": "",
        },
      },
    },
    "visData": Object {
      "columns": Array [
        Object {
          "id": "col-0-1",
          "name": "Count",
        },
      ],
      "rows": Array [
        Object {
          "col-0-1": 0,
        },
      ],
      "type": "datatable",
    },
    "visType": "metric",
  },
}
`;
